Svelte bundle size test

esheehan-gsl opened the pull request:
I wanted to compare bundle sizes between Preact and Svelte. Svelte is a little smaller, but not much (~2.5Kb). It’s hard to say whether this savings will become more significant as the application grows or not.

The primary motivation for moving to Preact from Svelte(Kit) was to avoid complicated build processes. We haven’t eliminated our build process entirely, and probably can’t because we will want to customize USWDS, which requires Sass. We were originally using SvelteKit, which has a much more complicated set up. The current implementation — which just uses Svelte without any of SvelteKit’s backend stuff — has a much simpler build process; in fact, it only required the addition of one plugin to our Vite config along with the Svelte compiler itself.

The components themselves are much simpler in Svelte than they were with Preact and Preact’s hooks, which bodes well for maintainability. Thinking about the long-term maintenance of this project, I think maintaining a Svelte application would be preferable to a Preact application, although the compiler piece makes me a little more nervous.

esheehan-gsl commented:
Part of what’s changed here for me regarding our choice of JS framework is that now that I’m moving over to ASCEND, the likelihood of continuing with this project beyond next spring is higher, and I’d much rather be maintaining a Svelte application than a Preact one.

ian-noaa commented:
I think that sounds reasonable. My recollection is we were nervous about relying on SvelteKit since it could change the bundled tooling on us between versions and - if this project were abandoned for a while - would require some work to get back to a point where we could begin to build the UI service. Svelte itself sounded like it was more stable than SvelteKit so I could see it being attractive if it'll simplify maintenance of the components. The changes to the build system here seem pretty straightforward.

We would continue to use web-components with Svelte in this setup?

esheehan-gsl commented:
My concerns over the build process still exist for this project, but the Vite setup we have now is simpler than what we had with SvelteKit, so it’s better for maintenance, I think. I had been looking into cutting out Vite entirely from the process after we switched to Preact/htm, but it I think that actually puts more onus on us to maintain the build tools, because we’d have to cobble together our bundler, Sass compiler, and dev server.

The other concern over SvelteKit was that it still hadn’t hit 1.0, so the API wasn’t, technically, stable. I’d been assured that in practice it was pretty stable and a 1.0 release was imminent in the next few months. It turned out, about a week after I got cold feet and switched to Preact, that they made a significant change to the API in SvelteKit. I don’t think it was to a part of the framework we were using, but it was a big disruption for many who’d been working SvelteKit. Svelte is on version 3, so it’s pretty stable at this point.

We would keep the web components, definitely. Moving to web components seems like a really good move for this project because it minimizes the amount of code we write in any given framework. So testing the move from Preact to Svelte only took a few hours because the web components that implement the charts didn’t change at all. The framework just acts as a glue for these components that helps manage application state and fetches data. And that’s good for long term maintenance of the application as well, since upgrading from, say, Svelte 3 to Svelte 4 would touch a lot less of the application, since the components wouldn’t have to change.
